[Feat/메인] 모든 게시글 가져오기

IToriginal opened the pull request:
## 🤔 Motivation
- 모든 게시글을 조회하는 기능을 개발 하였습니다.
- 사용자가 게시글의 조회를 최신순, 인기순으로 내용을 확인할 수 있도록 개발하였습니다.
- **최신순 조회** <br> 게시글을 생성한 순서대로(createdAt) 보여주어 사용자들이 최근에 작성된 게시글을 확인할 수 있습니다.
- **인기순 조회** <br> 좋아요 수(likeCnt)가 많은 게시글부터 보여주어 사용자들이 인기 있는 게시글을 우선순위로 확인할 수 있습니다.

<br>

## 💡 Key Changes
- `BoardRepository`에 `findAllByDeletedAtIsNullOrderByCreatedAtDesc()` 메서드를 추가하여 최신순으로 게시글을 조회할 수 있도록 하였습니다.
- `BoardRepository`에 `findAllByDeletedAtIsNullOrderByLikeCntDesc()` 메서드를 추가하여 인기순으로 게시글을 조회할 수 있도록 하였습니다.
- `BoardService` 인터페이스에 `readAllBoardsByLatest()` 메서드를 구현하여 최신순으로 게시글을 조회할 수 있도록 하였습니다.
- `BoardService` 인터페이스에 `readAllBoardsByPopularity()` 메서드를 구현하여 인기순으로 게시글을 조회할 수 있도록 하였습니다.
- `BoardController`에 최신순 조회와 인기순 조회에 해당하는 API를 추가하였습니다. <br> **API Path** <br>  -> 최신순 - `/`, 인기순 - `/popular`

<br>

## ✅ Test
- API 테스트는 PostMan을 통해서 진행했습니다.

**최신순**
<img width="1002" alt="image" src="https://github.com/four-uncles/saramara-community-server/assets/117193889/7e189bdb-529f-4c5d-bcdd-1e090f688ab4">

**인기순**
<img width="1002" alt="image" src="https://github.com/four-uncles/saramara-community-server/assets/117193889/9d3c807b-9b70-4591-95cb-65979a6f749b">

<br>

## 🧐 Insufficient Content
- 성능에 대한 고민이 있는데, 이 부분은 1차로 설계한 모든 API를 테스트해보면서 진행해볼 예정입니다.
예를 들어 메인에서 Attach의 이미지 파일을 불러오고, 게시글을 조회하게될 텐데, 두 API의 개발을 완료한 뒤에 약 100만의 더미데이터를 넣고 동시에 두개의 성능 테스트를 비교해보는게 어떨까 합니다. 아마 스프린트 5를 진행한 이후에 성능테스트 시간을 잡고 함께 비교해보면서 개선사항을 정리해 나가면 좋을 듯 합니다.

<br>

## 👨‍👨‍👧‍👦 To Reviewers
- @Jooney-95 @byeongJoo05 @hb9397 
- 위의 개발 내용에 대한 코드 리뷰 및 피드백 부탁드립니다.
- 추가적으로 개선할 수 있는 부분이나 주의해야 할 사항이 있다면 알려주시면 감사하겠습니다.

close: #40

IToriginal committed:
:sparkles: feat(#40): ResponseDTO 생성

IToriginal committed:
:sparkles: feat(#40): Repository 메서드 생성

게시글 전체를 조회하기 위한 method
- 1. 최신순
- 2. 인기순

IToriginal committed:
:sparkles: feat(#40): Service 로직 구현

- 최신순, 인기순은 정렬 방법만 틀리고 같은 값을 응답하기 때문에 Repository에서 생성한 정렬 메서드을 호출하고 공통적인 사항을 처리하기 위한 map.. 메서드로 반환하여 stream을 통해 list를 구현하였습니다.

IToriginal committed:
:sparkles: feat(#40): Controller 구현

- GetMapping
- 최신순: default 값으로 최신순을 사용하기 때문에 Path를 / 로 지정
- 인기순: 별도의 행위를 통해서 정렬된 값을 보여주기 위해서 API를 별도로 나눴습니다.

byeongJoo05 commented on src/main/java/com/kakao/saramaracommunity/board/repository/BoardRepository.java:
```diff
@@ -1,11 +1,19 @@
 
     Optional<Board> findByBoardIdAndDeletedAtIsNull(Long boardId);
+
+    @Query("SELECT b FROM Board b WHERE b.deletedAt IS NULL ORDER BY b.createdAt DESC")
+    List<Board> findAllByDeletedAtIsNullOrderByCreatedAtDesc();
```
`soft delete`로 사라지지 않은 데이터들은 제외하는 로직이 인상적입니다!

hb9397 replied:
왛 저는 
@Where(clause = "deleted_at is NULL")
@SQLDelete(sql = "update member set deleted_at = CURRENT_TIMESTAMP where member_id = ?")
@Entity
과 같이 Entity에 적용해서 하는 방식으로 검색에서 제외 되는 방식을 사용해 왔는데 조금 더 간결한 것 같습니다.

byeongJoo05 commented:
수고하셨습니다. 인기순/조회순을 한 api에서 처리하는 게 아닌 api 분리를 통하여 로직이 더욱 깔끔해보이네요!

langoustinee commented on src/main/java/com/kakao/saramaracommunity/board/controller/BoardController.java:
```diff
@@ -56,4 +57,33 @@ public ResponseEntity<Object> readOneBoard(@PathVariable Long boardId) {
         return ResponseEntity.ok(response);
     }
 
+    @GetMapping("/")
+    public ResponseEntity<Object> readAllBoardsByLatest() {
+        List<BoardResponseDto.ReadAllBoardResponseDto> boards = boardService.readAllBoardsByLatest();
+
+        // 응답 데이터 생성
+        Map<String, Object> response = new HashMap<>();
+        response.put("code", 200);
+        response.put("msg", "success");
+
+        // 게시글 목록
+        response.put("data", boards);
+
+        return ResponseEntity.ok(response);
+    }
+
+    @GetMapping("/popular")
+    public ResponseEntity<Object> readAllBoardsByPopularity() {
+        List<BoardResponseDto.ReadAllBoardResponseDto> boards = boardService.readAllBoardsByPopularity();
+
+        // 응답 데이터 생성
+        Map<String, Object> response = new HashMap<>();
+        response.put("code", 200);
+        response.put("msg", "success");
+
+        // 게시글 목록
+        response.put("data", boards);
+
+        return ResponseEntity.ok(response);
+    }
```
다양한 검색조건 별로 조회를 하는 API는 이번 미팅 시간에 하나의 API에서 구분할 것인지 API 자체를 구분할 것인지 이야기 해보면 좋을 것 같습니다.

langoustinee commented on src/main/java/com/kakao/saramaracommunity/board/repository/BoardRepository.java:
```diff
@@ -1,11 +1,19 @@
 
     Optional<Board> findByBoardIdAndDeletedAtIsNull(Long boardId);
+
+    @Query("SELECT b FROM Board b WHERE b.deletedAt IS NULL ORDER BY b.createdAt DESC")
+    List<Board> findAllByDeletedAtIsNullOrderByCreatedAtDesc();
```
저희 프로젝트 엔티티 생명주기 방식이 @byeongJoo05 님께서 말씀하셨듯이 Soft Delete 방식을 사용중입니다.

@hb9397 님께서 설명해주셨는데, 엔티티 클래스에 아래와 같은 구문을 추가로 작성한 것으로 알고 있습니다.

```java
@where(clause = "deleted_at is NULL")
@SQLDelete(sql = "update XXX set deleted_at = CURRENT_TIMESTAMP where XXX = ?")
public class XXX {
    ...
}
```

우리는 여기서  위 어노테이션의 `@Where`을 주의 깊게 봐야 합니다.

`@Where` 어노테이션은 해당 엔티티로 조회할 시, `deleted_at is NULL` 이 구문을 where절로 SQL에 적용시켜 줍니다.

헌욱님은 데이터 조회시 `WHERE b.deletedAt IS NULL`와 같은 옵션을 두셨는데 Board 엔티티 클래스에 `@Where` 어노테이션이 달려있지 않다면 해당 쿼리는 문제가 없습니다. 다만, 저희 프로젝트의 모든 엔티티에는 `@Where` 어노테이션이 달려있습니다.

그렇다면 헌욱님의 JPQL 쿼리에 무엇이 문제가 되고 있는지 잘 파악하실 수 있을 것이라 생각합니다!

IToriginal replied:
Entity 단에서 저희가 예전에 Null을 제외한 것을 잊고 코드를 작성해서 이 부분을 놓쳤네요. 저 대신 잘 찾아주셔서 감사합니다.
Meeting에서 의견을 나눴듯이 공통적으로 Entity에서 사용하고 있는 @Where이 있으니 JPA를 통해 Null을 한 번 더 찾을 필요는 없겠네요.
해당 부분은 수정해 두겠습니다!

langoustinee commented on src/main/java/com/kakao/saramaracommunity/board/service/BoardServiceImpl.java:
```diff
@@ -70,4 +72,43 @@ public BoardResponseDto.ReadOneBoardResponseDto readOneBoard(Long boardId) {
         return responseDto;
     }
+
+    @Override
+    public List<BoardResponseDto.ReadAllBoardResponseDto> readAllBoardsByLatest() {
+        List<Board> boards = boardRepository.findAllByDeletedAtIsNullOrderByCreatedAtDesc();
+
+        log.info("최신순으로 게시글을 조회합니다.(Reading all boards by latest)");
+
+        return mapBoardListToResponseDtoList(boards);
+    }
+
+    @Override
+    public List<BoardResponseDto.ReadAllBoardResponseDto> readAllBoardsByPopularity() {
+        List<Board> boards = boardRepository.findAllByDeletedAtIsNullOrderByLikeCntDesc();
+
+        log.info("인기순으로 게시글을 조회합니다.(Reading all boards by popularity)");
+
+        return mapBoardListToResponseDtoList(boards);
+    }
```
핵심 로직을 분리해서 메서드로 관리하니 가시성이 너무 좋은 것 같습니다..! 😀

IToriginal replied:
감사합니다. 하지만, API를 나눈다는게 좋을지는 고민되는 사항이라.. 어떤 방법을 사용할지에 따라서 해당 부분이 추후에 수정될 수 있을 듯 합니다!

IToriginal committed:
:hammer: fix(#40): CategoryBoard 컬럼명 변경

- Code Review 사항 반영

IToriginal committed:
:hammer: fix(#40): Repository 메서드 단순화

- Code Review 사항 반영
- Entity에서 이미 @Where 을 통해 Null인 사항을 필터링하고 있어서 Repository에서 하는 것은 비효율적인 상황이 발생했었다. 이 부분을 수정해서 메소드를 단순화 시킬수 있었다.
